### PR TITLE
enable CPU HBM

### PR DESCRIPTION
**Pull request by jikunshang**
this pr try to enable CPU HBM with memkind library to allocate hbm memory.

see https://github.com/ggerganov/llama.cpp/discussions/2602

### CODE REVIEW

**Comment by jikunshang** [comment]
> Are you seeing any performance improvements?

sorry for late reply, I just got test machine recently.
On Xeon CPU Max 9462,  model use llama2-7b:
baseline inference performance is 124 ms/token,
with this enhancement and HBM enabled, inference performance is 87 ms/token, which can get about 40% perf gain.

**Comment by jikunshang** [comment]
hi @ggerganov, can you take a review at your convenience?

File: ggml.c [hunk 1]
```diff
@@ -193,7 +196,13 @@ typedef void * thread_ret_t;
 #else
 inline static void * ggml_aligned_malloc(size_t size) {
     void * aligned_memory = NULL;
-#ifdef GGML_USE_METAL
+#ifdef GGML_USE_CPU_HBM
+    if (size == 0) {
+        GGML_PRINT("WARNING: Behavior may be unexpected when allocate 0 byte for hbw_posix_memalign!");
+        return NULL;
+    }
```
Comment (ggerganov) [inline ggml.c:203]: This check should be applied for all branches - it's always UB to allocate 0 bytes

**Review by ggerganov — CHANGES_REQUESTED** [review]

**Comment by hydroo** [comment]
Sorry for my ignorance.
Why doesn't the system allocate on HBM without this change?
Is this a system with DDR and HBM, and some BIOS setting (I vaguely remember caching vs other modes (For Optane and Knights* chips)) makes it such that the system either prioritizes DDR or even never touches HBM?
I'm asking, because it's unintuitive to me that a system wouldn't perhaps prioritize using HBM over DDR.

**Comment by jikunshang** [comment]
> Sorry for my ignorance. Why doesn't the system allocate on HBM without this change? Is this a system with DDR and HBM, and some BIOS setting (I vaguely remember caching vs other modes (For Optane and Knights* chips)) makes it such that the system either prioritizes DDR or even never touches HBM? I'm asking, because it's unintuitive to me that a system wouldn't perhaps prioritize using HBM over DDR.

Yes, you are right.
Actually, there are 3 kinds memory mode for Xeon Max Cpu serious: `HBM only memory mode`, `Flat memory mode(1LM)`, `Cache memory mode(2LM)`. For cache memory mode, it will work as you describe, but it will lock of fine grained memory management since all HBM memory are transparent,  like L4 cache. 
This code change is target for `Flat memory mode`, HBM and DDR are exposed to software as separate address space. we can use HBM on demand. 
More details about HBM configuration can be found [here](https://www.intel.com/content/www/us/en/content-details/782255/intel-xeon-cpu-max-series-configuration-and-tuning-guide.html)

File: ggml.c (hunk 1 above)
Comment (jikunshang) [reply ggml.c]: done.

File: ggml.c [hunk 2]
```diff
@@ -192,8 +195,14 @@ typedef void * thread_ret_t;
 #define GGML_ALIGNED_FREE(ptr)    _aligned_free(ptr)
 #else
 inline static void * ggml_aligned_malloc(size_t size) {
+    if (size == 0) {
+        GGML_PRINT("WARNING: Behavior may be unexpected when allocate 0 byte for ggml_aligned_malloc!");
```
Comment (ggerganov) [inline ggml.c:199]:
```suggestion
        GGML_PRINT("WARNING: Behavior may be unexpected when allocating 0 bytes for ggml_aligned_malloc!\n");
```

File: llama.cpp [hunk 3]
```diff
@@ -412,6 +415,9 @@ static void ggml_graph_compute_helper(std::vector<uint8_t> & buf, ggml_cgraph *
 #elif GGML_USE_METAL
 #   define llama_host_malloc(n)  ggml_metal_host_malloc(n)
 #   define llama_host_free(data) ggml_metal_host_free(data)
+#elif GGML_USE_CPU_HBM
+#   define llama_host_malloc(n)  hbw_malloc(n)
+#   define llama_host_free(data) if(data != NULL) hbw_free(data)
```
Comment (ggerganov) [inline llama.cpp:420]:
```suggestion
#   define llama_host_free(data) if (data != NULL) hbw_free(data)
```

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Merge if CI passes

**Comment by jikunshang** [comment]
Hi @ggerganov can you approve again for workflows? thanks!

**Comment by kunger97** [comment]
Does HBM compile the default in LLAMA.CPP or do you need to specify during compilation?

**Pull request merged**